### PR TITLE
Call `detectDatabasePlatform` only once

### PR DESCRIPTION
**Pull request by rosier**
Database platform detection is triggered twice if `Doctrine/DBAL/Connection::getDatabasePlatform()` is called before `Doctrine/DBAL/Connection::connect()`

### CODE REVIEW

**Comment by Ocramius** [comment]
@rosier requires a test case

File: lib/Doctrine/DBAL/Connection.php [hunk 1]
```diff
@@ -1597,7 +1593,7 @@ public function ping()
         }
 
         try {
-            $this->query($this->platform->getDummySelectSQL());
+            $this->query($this->getDatabasePlatform()->getDummySelectSQL());
```
Comment (deeky666) [inline lib/Doctrine/DBAL/Connection.php]: Why is this change necessary? `Connection::ping()` calls `Connection::connect()` as the beginning, so it ensures the platform is set.

Comment (rosier) [inline lib/Doctrine/DBAL/Connection.php]: The platform isn't set anymore in `connect()` with this PR.

In the current version of `Connection` the following calls are triggered:
`getDatabasePlatform()` => `detectDatabasePlatform()` => `getDatabasePlatformVersion()`  => `connect()` at this point because the platform isn't set yet `connect()` calls `detectDatabasePlatform()` again.

This PR removes the call to `detectDatabasePlatform()` from `connect()` to stop that from happening.

**Comment by rosier** [comment]
@Ocramius `Connection::detectDatabasePlatform()` is a private function. I'm not sure if there is a good way to test that it is called only once.

**Comment by Ocramius** [comment]
@rosier try looking at the wrapped driver: that one can be mocked

**Comment by deeky666** [comment]
Merged manually via https://github.com/doctrine/dbal/commit/de755610a46e7f3b91ad3070109364315e95fb73. Verified the issue by adding a test. Thanks @rosier !
